### PR TITLE
mgr: python interface rework + enable modules to run in standby mode

### PR DESCRIPTION
**Pull request by jcsp**
- [x] Update documentation + release notes for "mgr services" command.
- [x] Add mechanism for python modules to serve in standby mode

### CODE REVIEW

File: src/mgr/MgrStandby.cc [hunk 1]
```diff
@@ -172,13 +171,17 @@ void MgrStandby::send_beacon()
...
-  if (available && !available_in_map) {
-    // We are informing the mon that we are done initializing: inform
-    // it of our command set.  This has to happen after init() because
-    // it needs the python modules to have loaded.
-    m->set_command_descs(active_mgr->get_command_set());
-    dout(4) << "going active, including " << m->get_command_descs().size()
-            << " commands in beacon" << dendl;
+  if (available) {
+    if (!available_in_map) {
+      // We are informing the mon that we are done initializing: inform
```
Comment (amitkumar50) [inline src/mgr/MgrStandby.cc]: Multiple Line comment should be inside /* */

File: src/mon/MgrMap.h [hunk 2]
```diff
@@ -77,6 +77,10 @@ class MgrMap
   std::set<std::string> modules;
   std::set<std::string> available_modules;
 
+  // Map of module name to URI, indicating services exposed by
```
Comment (amitkumar50) [inline src/mon/MgrMap.h]: Multiple Line comment should be inside /* */

File: src/pybind/mgr/dashboard/module.py [hunk 3]
```diff
@@ -832,6 +832,13 @@ def get_perf_schema(self, **args):
             'engine.autoreload.on': False
         })
 
+        # Publish the URI that others may use to access the service we're
```
Comment (amitkumar50) [inline src/pybind/mgr/dashboard/module.py]: Use "we are" instead of we're as per coding guidelines. See
https://github.com/ceph/ceph/pull/16705/files

**Comment by wido** [comment]
Do we already want to include the "can_run()" method here? Like we discussed on the Ceph day. Meaning that we check if a module is able to run before we attempt it.

**Comment by jcsp** [comment]
@wido definitely want it, but this PR is already huge so let's do it separately afterwards

**Comment by wido** [comment]
@jcsp: Understood! Was scanning through the code and didn't see it. Just wanted to double-check I didn't miss anything.

File: src/pybind/mgr/dashboard/module.py [hunk 4]
```diff
@@ -184,6 +184,14 @@ def notify(self, notify_type, notify_val):
                     self.log_buffer.appendleft(notify_val)
         elif notify_type == "pg_summary":
             self.update_pool_stats()
+
+            #pg_summary = self.get("pg_summary")
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: this change is not related to the noop stub of `os._exit()`. can we drop it?

File: src/mgr/PyModuleRunner.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    assert(pClass != nullptr);
+    assert(pMyThreadState != nullptr);
+    assert(!module_name.empty());
+  }
+
+  ~PyModuleRunner();
+
+  PyModuleRunnerThread thread;
+
+  std::string const &get_name() { return module_name; }
```
Comment (tchaikov) [inline src/mgr/PyModuleRunner.h]: nit, make this method `const`.

File: src/mgr/ActivePyModules.cc [hunk 6]
```diff
@@ -324,218 +321,68 @@ PyObject *PyModules::get_python(const std::string &what)
...
+      module_name, pClass,
+      pMyThreadState));
 
-  dout(1) << "Creating threads for " << modules.size() << " modules" << dendl;
-  for (auto& i : modules) {
-    auto thread = new ServeThread(i.second.get());
-    serve_threads[i.first].reset(thread);
-  }
+  int r = modules[module_name]->load(this);
+  if (r != 0) {
```
Comment (tchaikov) [inline src/mgr/ActivePyModules.cc]: shall we remove the module failed to load from `modules`?

File: src/mgr/ActivePyModule.cc [hunk 7]
```diff
@@ -134,7 +134,19 @@ int ActivePyModule::load_commands()
   // Don't need a Gil here -- this is called from ActivePyModule::load(),
   // which already has one.
   PyObject *command_list = PyObject_GetAttrString(pClassInstance, "COMMANDS");
-  assert(command_list != nullptr);
+  if (command_list == nullptr) {
```
Comment (tchaikov) [inline src/mgr/ActivePyModule.cc]: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es87-dont-add-redundant--or--to-conditions

File: src/mgr/PyModuleRegistry.cc [hunk 8]
```diff
@@ -231,28 +231,29 @@ int PyModule::load(PyThreadState *pMainThreadState)
...
-    // Initialize base classes
-    BaseMgrModuleType.tp_new = PyType_GenericNew;
-    if (PyType_Ready(&BaseMgrModuleType) < 0) {
-        assert(0);
-    }
-
-    Py_INCREF(&BaseMgrModuleType);
-    PyModule_AddObject(ceph_module, "BaseMgrModule",
-                       (PyObject *)&BaseMgrModuleType);
+    auto load_class = [ceph_module](std::string name, PyTypeObject *type)
```
Comment (tchaikov) [inline src/mgr/PyModuleRegistry.cc]: nit, could just pass `const char*` instead of a `std::string`

File: src/mgr/PyModuleRegistry.cc [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+    PyObject *ceph_module = Py_InitModule("ceph_module", ModuleMethods);
+    assert(ceph_module != nullptr);
+
+    /*
+    Py_InitModule("ceph_osdmap", OSDMapMethods);
+    Py_InitModule("ceph_osdmap_incremental", OSDMapIncrementalMethods);
+    Py_InitModule("ceph_crushmap", CRUSHMapMethods);
+    */
+
+    auto load_class = [ceph_module](std::string name, PyTypeObject *type)
```
Comment (tchaikov) [inline src/mgr/PyModuleRegistry.cc]: nit, pass `const char*` instead.

**Comment by tchaikov** [comment]
@jcsp, could we split the mds related changes into separated PR(s)?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-10-30_14:46:06-rados-wip-sage-testing-2017-10-30-0722-distro-basic-smithi/

**Comment by liewegas** [comment]
What's left before we can merge this?  I was to get balancer merged!

File: src/mgr/PyModuleRegistry.cc [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+  // Drop the GIL and remember the main thread state (current
+  // thread state becomes NULL)
+  pMainThreadState = PyEval_SaveThread();
+
+  std::list<std::string> failed_modules;
+
+  // Load python code
+  for (const auto& module_name : mgr_map.modules) {
+    dout(1) << "Loading python module '" << module_name << "'" << dendl;
+    auto mod = std::unique_ptr<PyModule>(new PyModule(module_name));
```
[the remark this answers is not in the document]
Comment (jcsp) [reply src/mgr/PyModuleRegistry.cc]: done

File: src/mgr/PyModuleRunner.h (hunk 5 above)
Comment (jcsp) [reply src/mgr/PyModuleRunner.h]: done

File: src/pybind/mgr/dashboard/module.py (hunk 4 above)
Comment (jcsp) [reply src/pybind/mgr/dashboard/module.py]: oops!  done.

File: src/mgr/ActivePyModule.cc (hunk 7 above)
Comment (jcsp) [reply src/mgr/ActivePyModule.cc]: Hmm.  I kind of disagree with the guidelines on this one, I like being explicit that I'm checking a pointer rather than evaluating a class instance for truthiness.  I don't remember if the google C++ guide takes a position on this.

File: src/mgr/PyModuleRegistry.cc (hunk 9 above)
Comment (jcsp) [reply src/mgr/PyModuleRegistry.cc]: done

File: src/mgr/ActivePyModules.cc (hunk 6 above)
Comment (jcsp) [reply src/mgr/ActivePyModules.cc]: For the moment I suggest we keep them there -- later we might want to add a mechanism for reporting to the MgrMonitor about the list of modules that we tried and failed to load, so they need remembering somewhere.

**Comment by jcsp** [comment]
Those mds commits are gone (just pushed by accident while working on it).

**Comment by jcsp** [comment]
Added doc updates and one extra change to include disabled modules in the "module ls" output to make the enabling/disable workflow simpler.

**Comment by jcsp** [comment]
@liewegas this is good to go now.  No substantial changes since your last test run so I'd be inclined to reach for the big merge hammer...
